File: 2DVirtualWorldSimulatorCpp/2DVirtualWorldSimulator/Turtle.cpp
#include <iostream>
#include <cstdlib>
#include "Turtle.h"
#include "World.h"

Turtle::Turtle() {
	this->name = 'T';
	this->strength = 2;
	this->initiative = 1;
}

void Turtle::action(World* world) {
	//75% chance to stay in the same place.
	int side_to_go = rand() % 4;
	if (side_to_go == 0) {
		world->moveOrganism(this, 1);
	}
	else this->SetOrganismWantedPosition(this->position.x, this->position.y);
}

bool Turtle::collision(World* world,Organism* organismToFight) {
	//Reflects attacks of animal with strength less than 5. Attacker will return to the previous cell.
	
	if (this->isAttacker == true) {
		this->defaultAttack(world, organismToFight);
		return true;
	}
	else {
		if (organismToFight->GetStrength() < 5) {
			world->addCommentatorVector(ATTACK_REFLECTED, organismToFight, nullptr, NULL);
			return true;
		}
		else return false;
	}
}

char Turtle::GetName() const {
	return 'T';
}

string Turtle::GetOrganismFullName() const {
	return "Turtle";
}

Turtle::~Turtle() {

}

File: 2D_sym_python/animals/Turtle.py
from Animal import*
import random

class Turtle(Animal):

    def __init__(self):
        super().__init__()
        self.name = 'T'
        self.strength = 2
        self.initiative = 1
    
    def action(self, world):
        side_to_go = random.randint(0,3)
        if side_to_go==0:
            world.moveOrganism(self,1)
        else: self.SetOrganismWantedPosition(self.position['x'],self.position['y'])
    
    def collision(self, world, organismToFight):
        if self.isAttacker == True:
            self.defaultAttack(world, organismToFight)
            return True
        else:
            if organismToFight.GetStrength() < 5:
                world.AddCommentatorVector(reportType.ATTACK_REFLECTED,organismToFight,None,0)
                return True
            else: return False
    
    def GetName(self):
        return 'T'
    
    def GetOrganismFullName(self):
        return "Turtle"
    